test/bazel/proto/BUILD.bazel

```diff
@@ -17,3 +17,8 @@ typescript_proto_library(
     name = "pizza_service_ts_proto",
     proto = ":pizza_service_proto",
 )
+
+typescript_proto_library(
+    name = "external_ts_proto",
+    proto = "@external_protos//:external_proto",
```

Comment:
Why not use a well known proto, e.g. timestamp_proto? You wouldn't have to create a whole new repo then.

Reply:
I think I tried doing that but had trouble referencing them. I might not have tried timestamp though, and was looking for a shared proto in an already imported project. I can try with timestamp.

Reply:
This is the error I get when I try to swap out the external_proto for one of the protobuf ones:

```Bash
ERROR: file 'external/com_google_protobuf/type_pb_amd.js' is generated by these conflicting actions:
Label: @com_google_protobuf//:type_proto
RuleClass: proto_library rule
Configuration: b8c06cd9b0b7503287e86fef3f8d5e1f
Mnemonic: SkylarkAction
Action key: 56265e69de57015eea1415766535c1c8, f961eaf8efde02547c4e263567c12a69
Progress message: Creating AMD module for @com_google_protobuf//:type_proto
PrimaryInput: File:[[<execution_root>]bazel-out/darwin-fastbuild/bin]external/com_google_protobuf/type_pb.js
PrimaryOutput: File:[[<execution_root>]bazel-out/darwin-fastbuild/bin]external/com_google_protobuf/type_pb_amd.js
Primary outputs are different: 1005642992, 148794312
Owner information: []#//:defs.bzl%typescript_proto_library_aspect BuildConfigurationValue.Key[b8c06cd9b0b7503287e86fef3f8d5e1f] @com_google_protobuf//:type_proto BuildConfigurationValue.Key[b8c06cd9b0b7503287e86fef3f8d5e1f] false {}, []#@ts_protoc_gen//:defs.bzl%typescript_proto_library_aspect BuildConfigurationValue.Key[b8c06cd9b0b7503287e86fef3f8d5e1f] @com_google_protobuf//:type_proto BuildConfigurationValue.Key[b8c06cd9b0b7503287e86fef3f8d5e1f] false {}
MandatoryInputs: Attempted action contains artifacts not in previous action (first 5): 
        external/com_google_protobuf/type_pb.js
        src/bazel/convert_to_amd_bin.sh
        src/bazel/convert_to_amd
        _middlemen/src_Sbazel_Sconvert_Uto_Uamd-runfiles
Previous action contains artifacts not in attempted action (first 5): 
        external/com_google_protobuf/type_pb.js
        external/ts_protoc_gen/src/bazel/convert_to_amd_bin.sh
        external/ts_protoc_gen/src/bazel/convert_to_amd
        _middlemen/external_Sts_Uprotoc_Ugen_Ssrc_Sbazel_Sconvert_Uto_Uamd-runfiles
Outputs: Attempted action contains artifacts not in previous action (first 5): 
        external/com_google_protobuf/type_pb_amd.js
Previous action contains artifacts not in attempted action (first 5): 
        external/com_google_protobuf/type_pb_amd.js
ERROR: com.google.devtools.build.lib.actions.MutableActionGraph$ActionConflictException: for external/com_google_protobuf/type_pb_amd.js, previous action: action 'Creating AMD module for @co
m_google_protobuf//:type_proto', attempted action: action 'Creating AMD module for @com_google_protobuf//:type_proto'
INFO: Elapsed time: 0.333s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (1 packages loaded, 7 targets configured)
```

Reply:
I tried cloning you repo -- that error is caused by my previous change, the tool I added isn't referenced properly to work in external repos. In `defs.bzl`, it should be:

`//src/bazel:convert_to_amd` -> `"@ts_protoc_gen//src/bazel:convert_to_amd`

After changing that, I get another error about the files not being created. It looks like the external generated proto files are not being written to the correct directory. I ran out of time to debug tonight, but you can run: `bazel build -s --verbose_failures --sandbox_debug //test/bazel/proto:external_ts_proto` to see the command and look at the file structure.

Reply:
Ohhhh boy. Thanks for this. This makes it clear what the problem is. The files are being output to `bazel-bin/external/com_google_protobuf/google/protobuf/` but we're expecting them to be output to `bazel-bin/external/com_google_protobuf/`. That's because the _bazel_ package name is empty (`//:timestamp_proto`) but the declared package is [`google.protobuf`](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/timestamp.proto). So ironically, I was correct to put a check in to ensure the packages match. However, it's not the bazel package, it's the declared proto package, which I'm not sure if we can actually get access to in any way from this aspect (seems related to the same issue I'm running into [with Go](https://github.com/bazelbuild/rules_go/issues/707)). Plus, the check doesn't work when the bazel package is empty, so there's that, too...

Thoughts on how to resolve this?

Reply:
I took another look at this this morning and I think I've got it working.

There are two problems:

- Output files are not declared in the right directory
- The protoc output dir is not correct

For #-1, external files were being delcared as `external/com_google_protobuf/proto.proto` instead of `external/com_google_protobuf/google/protobuf/proto.proto`.

For #-2, the output dir should be the root of the repo, not the directory of the file. protoc handles writing to the sub dirs.

This reminded me of a similar [issue](https://github.com/pubref/rules_protobuf/pull/93) I fixed a few years ago. It may make sense to move this code into the rules_proto project so that we don't have to reinvent the wheel -- that's separate from this change though.

[Here are my changes](https://gist.github.com/Dig-Doug/5c971c2b3fe65046aabcc214677dbe9d), the output is a git diff. Can you see if they work for your use case? If so, feel free to patch them into this PR.

Reply:
Interesting. I was under the impression that protoc would output the files based on the package name, but it looks like it's done based on the path of the proto file. I'm not sure if that's desirable in the long-run, but it definitely makes this easier now. I see your solution makes sense now though - it just took me a while to figure out that it would generalize.

Reply:
For #1 you can just add `sibling=src` in `append_to_outputs` where it does `declare_file`, e.g.

```
output = ctx.actions.declare_file(file_name + ext, sibling=src)
```